app/helpers/concerns/foreman_teamdynamix/hosts_helper_extensions.rb

```diff
@@ -7,19 +7,47 @@ def teamdynamix_title
       SETTINGS[:teamdynamix][:title] || 'Team Dynamix'
     end
 
-    def teamdynamix_fields(host)
+    def teamdynamix_fields
       td_pane_fields = SETTINGS[:teamdynamix][:fields] || DEFAULT_TD_PANE_FIELDS
-      return [[_('Asset'), 'None Associated']] unless host.teamdynamix_asset_id
-      @asset = host.td_api.get_asset(host.teamdynamix_asset_id)
+      return [[_('Asset'), 'None Associated']] unless @host.teamdynamix_asset_id
+
+      get_teamdynamix_asset(@host.teamdynamix_asset_id)
+
+      # always display a link to the asset
+      fields = [ get_asset_uri ]
 
-      fields = []
       td_pane_fields.each do |field_name, asset_attr|
-        asset_attr_val = @asset[asset_attr]
-        fields += [[_(field_name.to_s), asset_attr_val]] if asset_attr_val.present?
+        asset_attr_val = @asset.has_key?(asset_attr) ? @asset[asset_attr] : get_nested_attrib_val(asset_attr)
+        fields += [[_(field_name.to_s), asset_attr_val]]
```

Comment:
We would still like the behavior of not adding a row for any field that is not present or empty on an asset. It should not blow up or raise an exception if the asset does not have the requested field. It should just silently not add that row. It is possible that some assets may have an attribute and others do not...

Does this code do that? IT seems like it blows up if the attribute is not found and if the attribute is blank it adds a blank row.

Reply:
In the case of nested attributes, if the parent attribute is not found that is probably because user may have mis-spelled it in configuration. If we silently fail that feedback is never provided to the user. We can find nested attributes within something that does not exist.
What I can do is support empty or nil values.

Reply:
Is it the case that all assets will have exactly the same nested fields? Or will some assets have a nested filed and another will not. This is what I am worried about. One hosts asset has an attribute or association to something else in TD and another hosts asset does not have that attribute or association.

Reply:
It will be obvious if the user misspelled a field, when they test it the filed will never show up. I think it is the lesser of two evils. This is the approach we took in the custom_tab plugin by using the try method here: https://github.com/MiamiOH/foreman_custom_tab/blob/0.1.0/app/helpers/concerns/foreman_custom_tab/hosts_helper_extensions.rb#L12

For example some hosts have facts and some do not b/c they don't run puppet.
Some hosts have certain associations that others do not.

I assume this could also be true in Teamdynamix. Some assets have associated tickets etc and some do not.

Reply:
I will make changes to address all comments related to attributes and nested attributes